raiko-block-downloader/src/provider/db.rs

```diff
@@ -0,0 +1,366 @@
+use alloy_primitives::{Address, Bytes, U256};
+use raiko_lib::{builder::OptimisticDatabase, consts::ChainSpec, mem_db::MemDb};
+use reth_primitives::{Header, B256};
+use reth_provider::ProviderError;
+use reth_revm::{
+    primitives::{Account, AccountInfo, Bytecode, HashMap},
+    Database, DatabaseCommit,
+};
+use std::{collections::HashSet, mem::take};
+use anyhow::Context;
+use tokio::runtime::Handle;
+
+use crate::{provider::BlockDataProvider, MerkleProof, RaikoResult};
+
+pub struct ProviderDb<BDP: BlockDataProvider> {
+    pub provider: BDP,
+    pub block_number: u64,
+    pub initial_db: MemDb,
+    pub initial_headers: HashMap<u64, Header>,
+    pub current_db: MemDb,
+    async_executor: Handle,
+
+    pub optimistic: bool,
+    pub staging_db: MemDb,
+    pub pending_accounts: HashSet<Address>,
+    pub pending_slots: HashSet<(Address, U256)>,
+    pub pending_block_hashes: HashSet<u64>,
+}
+
+impl<BDP: BlockDataProvider> ProviderDb<BDP> {
+    pub async fn new(provider: BDP, chain_spec: ChainSpec, block_number: u64) -> RaikoResult<Self> {
+        let mut provider_db = ProviderDb {
+            provider,
+            block_number,
+            async_executor: Handle::current(),
+            // defaults
+            optimistic: false,
+            staging_db: Default::default(),
+            initial_db: Default::default(),
+            initial_headers: Default::default(),
+            current_db: Default::default(),
+            pending_accounts: HashSet::new(),
+            pending_slots: HashSet::new(),
+            pending_block_hashes: HashSet::new(),
+        };
+        if chain_spec.is_taiko() {
+            // Get the 256 history block hashes from the provider at first time for anchor
+            // transaction.
+            let start = block_number.saturating_sub(255);
```

Comment:
the struct is generic over block provider trait while hardcodes `255` :) 
why is exactly `256` block hashes?